core/integration/module_up_test.go

```diff
@@ -13,17 +14,26 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// Note: ensure each testcase use unique port, otherwise you might see flakes
 func (ModuleSuite) TestDaggerUp(ctx context.Context, t *testctx.T) {
-	modDir := t.TempDir()
-	err := os.WriteFile(filepath.Join(modDir, "main.go"), []byte(`package main
+	// test both `container up` and `container as-service up`
+	for _, asObject := range []string{"container", "service"} {
+		defaultPort := "23400"
+
+		if asObject == "container" {
+			defaultPort = "23100"
+		}
```

Comment:
suggestions: use a testcase matrix pattern here so we don't need `if asObject == nil` throughout the tests. So essentially each testcase becomes:

```go
struct {
	name string
	port int
	cmd []string  // e.g. ["as-service", "up"]
}
```

Reply:
done. had to add some reusable function to make it work, but it looks nice now.